src/main/java/edu/kit/compiler/semantic/SemanticChecks.java

```diff
@@ -0,0 +1,253 @@
+import edu.kit.compiler.data.ast_nodes.StatementNode;
+import edu.kit.compiler.data.ast_nodes.ExpressionNode.ArrayAccessExpressionNode;
+import edu.kit.compiler.data.ast_nodes.ExpressionNode.BinaryExpressionNode;
+import edu.kit.compiler.data.ast_nodes.ExpressionNode.FieldAccessExpressionNode;
+import edu.kit.compiler.data.ast_nodes.ExpressionNode.IdentifierExpressionNode;
+import edu.kit.compiler.data.ast_nodes.ExpressionNode.MethodInvocationExpressionNode;
+import edu.kit.compiler.data.ast_nodes.ExpressionNode.NewArrayExpressionNode;
+import edu.kit.compiler.data.ast_nodes.ExpressionNode.NewObjectExpressionNode;
+import edu.kit.compiler.data.ast_nodes.ExpressionNode.ThisExpressionNode;
+import edu.kit.compiler.data.ast_nodes.ExpressionNode.UnaryExpressionNode;
+import edu.kit.compiler.data.ast_nodes.ExpressionNode.ValueExpressionNode;
+import edu.kit.compiler.data.ast_nodes.StatementNode.BlockStatementNode;
+import edu.kit.compiler.data.ast_nodes.StatementNode.ExpressionStatementNode;
+import edu.kit.compiler.data.ast_nodes.StatementNode.IfStatementNode;
+import edu.kit.compiler.data.ast_nodes.StatementNode.LocalVariableDeclarationStatementNode;
+import edu.kit.compiler.data.ast_nodes.StatementNode.ReturnStatementNode;
+import edu.kit.compiler.data.ast_nodes.StatementNode.WhileStatementNode;
+
+/**
+ * This class performs the remaining steps of semantic analysis after name and type analysis.
+ *
+ * Preconditions:
+ * - type checking is done
+ * - all references are initialized with the according definition
+ * - the definition for the builtin class "String" is available
+ *
+ * Postconditions:
+ * - left hand of assignments is an lvalue
+ * - all ExpressionStatements are either a method invocation or an assignment
+ * - if the return type of a method is not void, each branch contains a return statement
+ * - the class "String" is never constructed
+ * - in the main method, no parameters are accessed (args)
+ */
+public class SemanticChecks {
+    private static final DataType voidType = new DataType(DataTypeClass.Void);
+
+    public static void applyChecks(ProgramNode ast, ErrorHandler errorHandler, ClassNode stringClass) {
+        for (ClassNode currentClass: ast.getClasses()) {
+            for (MethodNode method: currentClass.getDynamicMethods()) {
+                MethodCheckVisitor visitor = new MethodCheckVisitor(false, errorHandler, stringClass);
+                boolean returns = method.getStatementBlock().accept(visitor);
+                if (!returns && !method.getType().equals(voidType)) {
+                    errorHandler.receive(new SemanticError(method,
+                        "return statement for all branches required"));
+                }
+            }
+            for (MethodNode method: currentClass.getStaticMethods()) {
+                MethodCheckVisitor visitor = new MethodCheckVisitor(true, errorHandler, stringClass);
+                method.getStatementBlock().accept(visitor);
+            }
```

Comment:
You could consider checking for `return` on all paths in static methods as well.